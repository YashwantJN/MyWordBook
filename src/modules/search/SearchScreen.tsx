import React, {useState} from 'react';
import {
  Modal,
  Pressable,
  ScrollView,
  StyleSheet,
  Text,
  TextInput,
  TouchableOpacity,
  View,
} from 'react-native';
import {NetworkManager} from '../../../networking/NetworkManager';
import PrimaryButton from '../../components/button/PrimaryButton';
import {Colors} from '../../constants/Colors.constant';
import NavigationItem from '../../components/items/NavigationItem';
import Loader from '../../components/loader/Loader';

const SearchScreen = ({navigation}) => {
  const [answer, onChangeAnswer] = useState('');
  const [wordNotFound, setWordNotFound] = useState(false);
  const [wordsList, setWordsList] = useState([]);
  const [searching, setSearching] = useState(false);

  function navigateToInfo(word:any) {
    console.log('wordData', word);
    
    navigation.navigate('Info', {info: {
      word: word.word,
      definition: word?.meanings[0]?.definitions[0]?.definition,
      phonetics: word.phonetics[0]?.audio,
    }})
  }

  function onChangeText(enteredText: string): void {
    onChangeAnswer(enteredText);
  }

  async function onPressSearch() {
    onChangeAnswer('');
    setWordsList([]);
    setSearching(true)
    await NetworkManager(
      (response: any) => {
        setSearching(false)
        if (response === 'not found') {
          setWordNotFound(true);
        } else {
          setWordsList(response);
        }
      },
      `/${answer}`,
      'GET',
    );
  }

  return (
    <View style={styles.container}>
      <Modal transparent visible={wordNotFound} animationType="slide">
        <View
          style={{
            width: '100%',
            height: '30%',
            justifyContent: 'center',
            alignItems: 'center',
            backgroundColor: Colors.button,
          }}>
          <Text style={{color: 'black', fontSize: 25, padding: 5}}>
            Not Found
          </Text>
          <Text
            style={{
              color: 'black',
              fontSize: 10,
              fontWeight: '200',
              padding: 5,
            }}>
            Please enter valid word
          </Text>
          <TouchableOpacity
            style={{
              backgroundColor: 'red',
              paddingHorizontal: 75,
              paddingVertical: 10,
              borderRadius: 5,
            }}
            onPressOut={() => {
              setWordNotFound(false);
            }}>
            <Text>Close</Text>
          </TouchableOpacity>
        </View>
      </Modal>
      <View>
        <TextInput
          style={styles.input}
          onChangeText={onChangeText}
          value={answer}
          placeholder="Search for a word..."
          placeholderTextColor="gray"
          keyboardType="default"
          autoCapitalize="none"
          autoCorrect={false}
          editable={true}
          selectTextOnFocus={false}
          autoComplete="off"
          maxLength={15}
        />
        <PrimaryButton title="Search" onPress={onPressSearch} />
      </View>
      <ScrollView>
        {wordsList.map((word: any) => {
          return (
            <View key={Math.random()}>
              <NavigationItem
                title={word?.word}
                description={word?.meanings[0]?.definitions[0]?.definition}
                onPress={() => {
                  navigateToInfo(word)
                }}
              />
            </View>
          );
        })}
      </ScrollView>
      <Loader isLoading={searching} text='Searching...'/>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    padding: 16,
  },
  searchView: {
    flex: 1,
    flexDirection: 'row',
  },
  input: {
    maxHeight: '40%',
    height: 40,
    borderWidth: 2,
    borderRadius: 5,
    paddingVertical: 2,
    paddingHorizontal: 10,
    borderColor: Colors.border,
  },
});

export default SearchScreen;
