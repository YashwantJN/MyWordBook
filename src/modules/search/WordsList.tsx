import React from 'react';
import {ScrollView, View} from 'react-native';
import NavigationItem from '../../components/items/NavigationItem';

const WordsList = ({wordsList, navigateToInfo}) => {
  return (
    <ScrollView>
      {wordsList?.map((word: any) => {
        return (
          <View key={Math.random()}>
            <NavigationItem
              title={word?.word}
              description={word?.meanings[0]?.definitions[0]?.definition}
              onPress={() => {
                navigateToInfo(word);
              }}
            />
          </View>
        );
      })}
    </ScrollView>
  );
};

export default WordsList;
