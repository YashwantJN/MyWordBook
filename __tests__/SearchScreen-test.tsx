/**
 * @format
 */

import 'react-native';
import React from 'react';
import renderer, {act} from 'react-test-renderer';
import {fireEvent, render, waitFor} from '@testing-library/react-native';
import SearchScreen from '../src/modules/search/SearchScreen';
import WordsList from '../src/modules/search/WordsList';

describe('SearchScreen', () => {
  test('search screen renders correctly', async () => {
    const navigation = jest.fn();
    const tree = renderer
      .create(<SearchScreen navigation={navigation} />)
      .toJSON();
    await act(async () => {
      expect(tree).toMatchSnapshot();
    });
  });

  test('should update input value when typing', async () => {
    const mockNavigate = jest.fn();

    const {getByPlaceholderText} = render(
      <SearchScreen navigation={{navigate: mockNavigate}} />,
    );

    const input = getByPlaceholderText('Search for a word...');
    fireEvent.changeText(input, 'Hello');

    expect(input.props.value).toBe('Hello');
  });

  test('searched value should display in child component', async () => {
    const mockNavigateToInfo = jest.fn();
    const mockWordsList = [
      {
        word: 'hello',
        phonetics: [
          {
            audio:
              '//ssl.gstatic.com/dictionary/static/sounds/20200429/hello--_gb_1.mp3',
          },
        ],
        meanings: [
          {
            definitions: [
              {
                definition:
                  'used as a greeting or to begin a phone conversation.',
              },
            ],
          },
        ],
      },
    ];

    const {queryByText, getAllByTestId} = render(
      <WordsList
        wordsList={mockWordsList}
        navigateToInfo={mockNavigateToInfo}
      />,
    );

    const items = getAllByTestId('NavigationItem');
    fireEvent.press(items[0]);

    await waitFor(() => {
      expect(queryByText('hello')).toBeTruthy();
      expect(mockNavigateToInfo).toHaveBeenCalledTimes(1);
    });
  });
});

// const {getByPlaceholderText, getAllByTestId} = render(
//     <SearchScreen navigation={{navigate: mockNavigateToInfo}} />,
//   );

//   const input = getByPlaceholderText('Search for a word...');
//   fireEvent.changeText(input, 'Hello');

//   expect(input.props.value).toBe('Hello');

//   const search = getAllByTestId('primaryButton');
//   fireEvent.press(search[0]);
