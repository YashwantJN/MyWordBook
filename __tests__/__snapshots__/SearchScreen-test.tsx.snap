// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchScreen search screen renders correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "padding": 16,
    }
  }
>
  <Modal
    animationType="slide"
    hardwareAccelerated={false}
    transparent={true}
    visible={false}
  />
  <View>
    <TextInput
      autoCapitalize="none"
      autoComplete="off"
      autoCorrect={false}
      editable={true}
      keyboardType="default"
      maxLength={15}
      onChangeText={[Function]}
      placeholder="Search for a word..."
      placeholderTextColor="gray"
      selectTextOnFocus={false}
      style={
        Object {
          "borderColor": "#aac8a7",
          "borderRadius": 5,
          "borderWidth": 2,
          "height": 40,
          "maxHeight": "40%",
          "paddingHorizontal": 10,
          "paddingVertical": 2,
        }
      }
      value=""
    />
    <View
      accessibilityState={
        Object {
          "disabled": false,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "alignSelf": "center",
          "backgroundColor": "#ffd966",
          "borderColor": "#aac8a7",
          "borderRadius": 36,
          "borderWidth": 2,
          "justifyContent": "center",
          "marginTop": 24,
          "opacity": 1,
          "paddingHorizontal": 32,
          "paddingVertical": 14,
          "width": "100%",
        }
      }
      testID="primaryButton"
    >
      <Text
        style={
          Object {
            "color": "#040d12",
            "fontSize": 18,
            "fontWeight": "600",
            "letterSpacing": 2,
            "textAlign": "center",
          }
        }
      >
        Search
      </Text>
    </View>
  </View>
  <RCTScrollView>
    <View />
  </RCTScrollView>
  <Modal
    animationType="none"
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    statusBarTranslucent={true}
    style={
      Object {
        "zIndex": 1100,
      }
    }
    transparent={true}
    visible={false}
  />
</View>
`;
